contracts/tokens/SecurityToken.sol

```diff
@@ -299,12 +310,70 @@ contract SecurityToken is ISecurityToken {
         emit LogFreezeTransfers(freeze, now);
     }
 
+    function adjustTotalSupplyCheckpoints() internal {
+      //No checkpoints set yet
+      if (currentCheckpointId == 0) {
+          return;
+      }
+      //No previous checkpoint data - add current balance against checkpoint
+      if (checkpointTotalSupply.length == 0) {
+          checkpointTotalSupply.push(
+              Checkpoint({
+                  checkpointId: currentCheckpointId,
+                  value: totalSupply()
+              })
+          );
+          return;
+      }
+      //No new checkpoints since last update
+      if (checkpointTotalSupply[checkpointTotalSupply.length - 1].checkpointId == currentCheckpointId) {
+          return;
+      }
+      //New checkpoint, so record balance
+      checkpointTotalSupply.push(
+          Checkpoint({
+              checkpointId: currentCheckpointId,
+              value: totalSupply()
+          })
+      );
+    }
+
+    function adjustBalanceCheckpoints(address _investor) internal {
+        //No checkpoints set yet
+        if (currentCheckpointId == 0) {
+            return;
+        }
+        //No previous checkpoint data - add current balance against checkpoint
+        if (checkpointBalances[_investor].length == 0) {
+            checkpointBalances[_investor].push(
+                Checkpoint({
+                    checkpointId: currentCheckpointId,
+                    value: balanceOf(_investor)
+                })
+            );
+            return;
+        }
+        //No new checkpoints since last update
+        if (checkpointBalances[_investor][checkpointBalances[_investor].length - 1].checkpointId == currentCheckpointId) {
```

Comment:
This line only allows a SINGLE checkpointBalance to be created per checkpoint. If on checkpoint 1 account 0xabc makes 2 transfers, only the first one will be "recorded". Is that correct?

Reply:
That's correct - the first time a transfer is made after a given checkpoint, the balance *before* the transfer is written out for the checkpoint, so that the balance written out corresponds to the time at which the checkpoint was made.